keps/sig-release/20190118-creating-packages.md

```diff
@@ -0,0 +1,417 @@
+Freeze](https://github.com/kubernetes/sig-release/tree/master/releases) of the
+targeted release**.
+-->
+
+## Release Signoff Checklist
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] k/enhancements issue in release milestone and linked to KEP (insert link here)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+
+## Summary
+
+There are currently different ways to build packages (`deb`s and
+`rpm`s) for the kubernetes components. All of which have some shortcomings and
+it is hard to understand for users which one is the correct one.
+The goal is to streamline all the package building mechanism, create one way on
+how to generate packages, and retire all the other options.
+
+## Motivation
+
+As mentioned in the summary, there are multiple different ways on how to create
+packages for kubernetes:
+
+- [`k/release/debian`][k-release-deb] & [`k/release/rpm`][k-release-rpm]
+  - *Description*:  
+    These tools are part of the official release process. After a new
+    kubernetes version has been cut and released, but before the new release
+    was announced, these tools will pull the build binaries from a bucket and
+    wrap them in a `deb`/`rpm`.
+  - *Usage*:
+    ```
+    k/release # ./debian/jenkins.sh
+    k/release # cd ./rpm/ && ./docker-build.sh
+    ```
+  - *Cons*<a id="k-release-cons"></a>:
+    - It uses a different (no) branching strategy then the artefacts build from `k/k` it is packaging
+    - Pulls pre-existing artefacts from a GCS bucket
+    - Cannot build just one package, builds all the packages for all architectures (on linux)
+    - Needs manual change of the source code when bumping versions
+  - *Pros*:
+    - builds for multiple platforms (`linux/aarch64`, `linux/armhfp`, `linux/ppc64le`, `linux/s390x`, `linux/x86_64`)
+    - also handles the upload of the packages (at least for `deb`s)
+      - might also be considered a [con](#k-release-cons)?
+  - *Parts to reuse*:
+    - multi-arch support
+- [`k/k/debs`][k-k-deb] & [`k/k/rpms`][k-k-rpm]
+  - *Description*:  
+    This is building the packages directly with bazel. Those packages are not
+    used as official releases -- usecase is not clear.
+  - *Usage*:
+    ```
+    k/k # bazel build //build/debs
+    k/k # bazel build //build/rpms
+    ```
+  - *Cons*:
+    - No multi-arch support, builds only `linux/amd64`
+    - rpm packaging needs `rpmbuild` to be installed on the host
```

Comment:
this is roughly as true for for the k/release build, FWIW, both would require it but for both we have access to docker images that include dependencies like this.

Reply:
What I mean is that on my dev machine
- running `bazel build //build/rpms` in `k/k` requires me to have `rpmbuild` and GNU `install` installed
- running `cd rpm && ./docker-build.sh` in `k/release` "just works"

Ideally, I'd love to see my package build(s) to "just work", regardless of where I do them. And as you said, there are docker images around which have those deps packaged anyway, as an example the one that is used for the published packages here [`k/release/rpm/Dockerfile`](https://github.com/kubernetes/release/blob/master/rpm/Dockerfile).

Reply:
This depends on the build tooling.  Most of the tools use automated container build tools to generate the artifacts.

Reply:
@hoegaarden bazel is also containerizeable with rpmbuild and we have multiple images that do that including [a local wrapper](https://github.com/kubernetes/test-infra/tree/master/planter), the only reason the k/release rpm build doesn't need one is because we wrap a similar container with rpmbuild in it.

I don't think there's much of a distinction there. We can easily ship a container for pretty much whatever tool we choose.

Reply:
FWIW it's feasible for bazel to automatically download an `rpmbuild` binary, rather than requiring it to be installed on the host system, but the `pkg_rpm` rules don't really have an active maintainer AFAICT.